zebra/zapi_msg.c

```diff
@@ -2755,9 +2755,9 @@ static inline void zread_iptable(ZAPI_HANDLER_ARGS)
 		zebra_pbr_add_iptable(zpi);
 	else
 		zebra_pbr_del_iptable(zpi);
-
+	return;
```

Comment:
`zebra_pbr_add_iptable` doesn't store a reference to `zpi`.
`zebra_pbr_del_iptable` doesn't free `zpi`.

What are you fixing here?

Reply:
zebra_pbr_iptable_free() calls zebra_pbr_add_iptable(), and this is my concern.

Reply:
Please post the line where `zebra_pbr_iptable_free` calls `zebra_pbr_add_iptable`.

Reply:
zebra_pbr_iptable_free()  deals only with freeing memory, not detach from list

Reply:
there was however a hook called, which totally changes the underlying behaviour

Reply:
The issue can be seen on master when testing with a plugin attached to zebra_pbr_table.
If you dont have plugin, the issue will not appear.

Actually, the zapi call() is supposed to trigger a call to that plugin.

    ADD_TABLE will call plugin_handle_iptable(ctxt, 1)

if we could set up a trace on the plugin handler, we could observe the following:

plugin_handle_iptable(ctxt, 1)
plugin_handle_iptable(ctxt, 0)

actually, the second call is done by the function that is in charge of freeing the iptable context allocated by zapi. the function in charge of freeing the iptable context is currently used to free both the zapi context and the entries in the hash list of pbr contexts.

as it is common, freeing zapi context was also reverting the plugin call.

Reply:
@qlyoung , any feedbacks please?